clap_generate/src/man.rs

```diff
@@ -0,0 +1,287 @@
+use std::fmt::write;
+
+use clap::ArgSettings;
+
+/// Man page generator
+#[derive(Debug, Clone)]
+pub struct Man<'a> {
+    app: clap::App<'a>,
+    section: i8,
+    manual: Option<String>,
+    sections: Vec<(String, Vec<String>)>,
+}
+
+impl<'a> Man<'a> {
+    /// Create a new man page for the app.
+    pub fn create(
+        app: clap::App<'a>,
+        section: i8,
+        manual: Option<String>,
+        buf: &mut dyn std::io::Write,
+    ) {
+        let man = Man {
+            app,
+            section,
+            manual,
+            sections: Vec::new(),
+        };
+
+        man.write(buf)
+    }
+
+    /// Create a new builder for man pages.
+    pub fn new(app: clap::App<'a>, section: i8, manual: Option<String>) -> Self {
+        Man {
+            app,
+            section,
+            manual,
+            sections: Vec::new(),
+        }
+    }
+
+    /// Add a custom section to the man pages.
+    pub fn section(mut self, title: impl Into<String>, body: Vec<impl Into<String>>) -> Self {
+        self.sections
+            .push((title.into(), body.into_iter().map(|s| s.into()).collect()));
+        self
+    }
+
+    /// Write the manpage to a buffer.
+    pub fn write(self, buf: &mut dyn std::io::Write) {
+        let mut buffer = String::new();
+
+        self.render_header(&mut buffer);
+
+        // Set sentence_space_size to 0 to prevent extra space between sentences separated
+        // by a newline the alternative is to add \& at the end of the line
+        buffer.push_str(".ss \\n[.ss] 0\n");
+        // Disable hyphenation
+        buffer.push_str(".nh\n");
+        // Disable justification (adjust text to the left margin only)
+        buffer.push_str(".ad l\n");
+
+        self.render_about(&mut buffer);
+        self.render_description(&mut buffer);
+        self.render_synopsis(&mut buffer);
+
+        self.render_options(&mut buffer);
+        self.render_subcommands(&mut buffer);
+
+        self.render_custom_sections(&mut buffer);
+
+        self.render_version(&mut buffer);
+        self.render_authors(&mut buffer);
+
+        w!(buf, buffer.as_bytes());
+    }
+
+    fn render_header(&self, buf: &mut String) {
+        write(
+            buf,
+            format_args!(
+                ".TH {} {}",
```

Comment:
@sondr3 in a [comment](https://github.com/clap-rs/clap/pull/3174#issuecomment-993891125)
> As for using libraries, I considered it but didn't want to be dependent on them and I took "inspiration" from clap_generate being dependency-free (besides depending on clap).

Reply:
@epage  in a [comment](https://github.com/clap-rs/clap/pull/3174#issuecomment-993901785)

> I can understand. The issue of how much to read into precedence or not when starting a contribution. iirc roff was originally written with clap man page generation in mind. The rest of clap_generate doesn't have dependencies because nearly all of it is generating for arbitrary programming languages rather than a well defined markup. I can see has added shlex at some point if we need to be fixing quoting and string joining problems.

Reply:
@sondr3 in a [comment](https://github.com/clap-rs/clap/pull/3174#issuecomment-994789904)
> I did not end up using the roff crate, mostly because I had what I needed already though I'm not against using it. I

Reply:
@epage I can switch to using `roff`/`man`, but I'm worried about the lack of maintenance for those crates and they are missing a fair bit of functionality last I looked, I see you're also active in the `rust-cli` repositories so if you'd be willing to I can send contributions there and incorporate them back into this :smile:

Reply:
To be clear, I was just centralizing the conversation as I consider this

When dealing with structured data, I tend to prefer creating a dedicated API for it, complete with tests.  Depending on the maturity, I might keep it in the same crate, but move towards splitting it out as it matures which I generally find helps it mature more.

You've done a step in that direction with `man.rs` though it could use some testing (e.g. `escape` is a prime candidate for testing)

I lean towards us reviving `roff` (or using the other one) but I'm mixed on where to prioritize that.

Reply:
> I'll look into reviving the roff crate.

Depending on how far along you are, I think i'm leaning towards let's get this crate merged and iterate on it from there.  There will always be more improvements.

Reply:
Sorry, I didn't notice this comment when I started working on converting this to using my fork of `roff-rs`, I've just pushed a full refactor for this to use it :see_no_evil:

Reply:
Can you create a PR against the roff repo so I can start reviewing those changes and get a release out?

Reply:
I've updated CI on the roff repo.  I did make some changes to make rustfmt / clippy happy. Feel free to overwrite those with your changes.

Reply:
I've opened a PR for my changes at rust-cli/roff-rs#14. I had also made some changes to the CI but yours is far more robust, so I've removed my changes.